src/libConsensus/ConsensusLeader.cpp

```diff
@@ -63,9 +63,207 @@ bool ConsensusLeader::CheckState(Action action) {
   return true;
 }
 
+bool ConsensusLeader::CheckStateSubset(uint16_t subsetID, Action action) {
+  ConsensusSubset& subset = m_consensusSubsets.at(subsetID);
+
+  static const std::multimap<ConsensusCommon::State, Action> ACTIONS_FOR_STATE =
+      {{CHALLENGE_DONE, PROCESS_RESPONSE},
+       {FINALCHALLENGE_DONE, PROCESS_FINALRESPONSE}};
+
+  bool found = false;
+
+  for (auto pos = ACTIONS_FOR_STATE.lower_bound(subset.m_state);
+       pos != ACTIONS_FOR_STATE.upper_bound(subset.m_state); pos++) {
+    if (pos->second == action) {
+      found = true;
+      break;
+    }
+  }
+
+  if (!found) {
+    LOG_GENERAL(WARNING, "SubsetID: " << subsetID << ", Action "
+                                      << GetActionString(action)
+                                      << " not allowed in state "
+                                      << GetStateString());
+    return false;
+  }
+
+  return true;
+}
+
+void ConsensusLeader::SetStateSubset(uint16_t subsetID, State newState) {
+  LOG_MARKER();
+  if ((newState == INITIAL) ||
+      (newState > m_consensusSubsets.at(subsetID).m_state)) {
+    m_consensusSubsets.at(subsetID).m_state = newState;
+  }
+}
+
+void ConsensusLeader::GenerateConsensusSubsets() {
+  LOG_MARKER();
+  lock_guard<mutex> g(m_mutex);
+
+  // Get the list of all the peers who committed, by peer index
+  vector<unsigned int> peersWhoCommitted;
+  for (unsigned int index = 0; index < m_commitMap.size(); index++) {
+    if (m_commitMap.at(index) && index != m_myID) {
+      peersWhoCommitted.push_back(index);
+    }
+  }
+  // Generate NUM_CONSENSUS_SUBSETS lists (= subsets of peersWhoCommitted)
+  // If we have exactly the minimum num required for consensus, no point making
+  // more than 1 subset
+
+  const unsigned int numSubsets = (peersWhoCommitted.size() < m_numForConsensus)
```

Comment:
If `peersWhoCommitted.size`  is less than the required minimum for consensus, should we still proceed with the protocol?

Reply:
No. Good catch. Will add a check.

Reply:
Ah actually your check for `(m_commitCounter < m_numForConsensus)` in the timed functions below already check for it, so `(peersWhoCommitted.size() < m_numForConsensus)` should never happen.  I think you can change the `<` to `==`.

Reply:
Now I remember, why I had used < instead of ==. If I had used ==,  I have to put this ‘peerwhocommitted.size() == m_numForConsus - 1’ becoz peerswhocommitted excludes commit by myself. And to avoid additional arithmetic op for - 1, I had stick to ‘peerswhocommitted.size() < m_numForConsensus’. So I think current way should be ok.